File: backend/app/schemas/participant.py
import enum
import re
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator


class AttendanceStatus(str, enum.Enum):
    REGISTERED = "REGISTERED"
    PRESENT = "PRESENT"
    ABSENT = "ABSENT"


class RegistrationSource(str, enum.Enum):
    GOOGLE_FORM = "GOOGLE_FORM"


def clean_text(value: str | None) -> str | None:
    return " ".join(value.split()) if value is not None else None


class ParticipantRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    full_name: str
    email: EmailStr
    phone: str | None
    college: str | None
    department: str | None
    year: str | None
    team_name: str | None
    registration_source: RegistrationSource
    registration_timestamp: datetime
    attendance_status: AttendanceStatus
    certificate_eligible: bool
    is_disqualified: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime


class ParticipantUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    full_name: str | None = Field(default=None, min_length=1, max_length=160)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=32)
    college: str | None = Field(default=None, max_length=160)
    department: str | None = Field(default=None, max_length=160)
    year: str | None = Field(default=None, max_length=32)
    team_name: str | None = Field(default=None, max_length=160)
    attendance_status: AttendanceStatus | None = None
    certificate_eligible: bool | None = None
    is_disqualified: bool | None = None
    notes: str | None = Field(default=None, max_length=5000)

    @field_validator("full_name", "phone", "college", "department", "year", "team_name", mode="before")
    @classmethod
    def normalize_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not isinstance(value, str):
            raise ValueError("Must be text")
        return clean_text(value)

    @field_validator("email")
    @classmethod
    def normalize_email(cls, value: EmailStr | None) -> str | None:
        return str(value).lower() if value is not None else None

    @field_validator("phone")
    @classmethod
    def valid_phone(cls, value: str | None) -> str | None:
        if value and not re.fullmatch(r"[+0-9 ()-]{7,32}", value):
            raise ValueError("Invalid phone number")
        return value or None


class BulkAction(str, enum.Enum):
    MARK_PRESENT = "MARK_PRESENT"
    MARK_ABSENT = "MARK_ABSENT"
    MARK_ELIGIBLE = "MARK_ELIGIBLE"
    MARK_INELIGIBLE = "MARK_INELIGIBLE"


class BulkUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    participant_ids: list[int] = Field(min_length=1, max_length=500)
    action: BulkAction

    @field_validator("participant_ids")
    @classmethod
    def unique_ids(cls, value: list[int]) -> list[int]:
        if any(item <= 0 for item in value):
            raise ValueError("Participant IDs must be positive")
        return list(dict.fromkeys(value))


class ParticipantFilters(BaseModel):
    search: str | None = Field(default=None, max_length=160)
    attendance: AttendanceStatus | None = None
    certificate_eligible: bool | None = None
    college: str | None = Field(default=None, max_length=160)
    year: str | None = Field(default=None, max_length=32)
    is_disqualified: bool | None = None
    sort_by: Literal["name", "registration_date", "college", "team"] = "registration_date"
    sort_dir: Literal["asc", "desc"] = "desc"


class ParticipantPage(BaseModel):
    items: list[ParticipantRead]
    page: int
    page_size: int
    total: int
    total_pages: int


class BulkResult(BaseModel):
    updated_count: int

File: frontend/src/types/participant.ts
export type AttendanceStatus = 'REGISTERED' | 'PRESENT' | 'ABSENT'
export interface CsvParticipant { id: number; name: string; email: string; team_name: string; college: string; created_at: string }
export interface CsvParticipantPage { items: CsvParticipant[]; page: number; page_size: number; total: number; total_pages: number }
export interface CsvPreviewRow { line: number; name: string; email: string; team_name: string; college: string; status: 'VALID' | 'DUPLICATE' | 'INVALID'; reason: string }
export interface CsvPreview { digest: string; rows: CsvPreviewRow[]; valid: number; duplicates: number; invalid: number }
export interface CsvImportResult { imported: number; duplicates: number; invalid: number }
export type BulkAction = 'MARK_PRESENT' | 'MARK_ABSENT' | 'MARK_ELIGIBLE' | 'MARK_INELIGIBLE'

export interface Participant {
  id: number
  full_name: string
  email: string
  phone: string | null
  college: string | null
  department: string | null
  year: string | null
  team_name: string | null
  registration_source: 'GOOGLE_FORM'
  registration_timestamp: string
  attendance_status: AttendanceStatus
  certificate_eligible: boolean
  is_disqualified: boolean
  notes: string | null
  created_at: string
  updated_at: string
}

export interface ParticipantPage {
  items: Participant[]
  page: number
  page_size: number
  total: number
  total_pages: number
}

export interface ParticipantFilters {
  search?: string
  attendance?: AttendanceStatus | ''
  certificate_eligible?: 'true' | 'false' | ''
  college?: string
  year?: string
  is_disqualified?: 'true' | 'false' | ''
  sort_by?: 'name' | 'registration_date' | 'college' | 'team'
  sort_dir?: 'asc' | 'desc'
}

export interface FilterOptions {
  colleges: string[]
  years: string[]
}

export type ParticipantChanges = Partial<Pick<Participant, 'full_name' | 'email' | 'phone' | 'college' | 'department' | 'year' | 'attendance_status' | 'certificate_eligible' | 'is_disqualified' | 'notes'>> & { team_name?: string | null }
